agent/hcp/telemetry/otel_sink.go

```diff
@@ -0,0 +1,198 @@
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/hcp/client"
+	"github.com/hashicorp/go-hclog"
+
+	gometrics "github.com/armon/go-metrics"
+	"go.opentelemetry.io/otel/attribute"
+	otelmetric "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/instrument"
+	otelsdk "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+type OTELSinkOpts struct {
+	Reader otelsdk.Reader
+	Logger hclog.Logger
+	Ctx    context.Context
+}
+
+type OTELSink struct {
+	spaceReplacer *strings.Replacer
+	logger        hclog.Logger
+	ctx           context.Context
+
+	meterProvider *otelsdk.MeterProvider
+	meter         *otelmetric.Meter
+
+	gaugeInstruments     map[string]*instrument.Float64ObservableGauge
```

Comment:
With sync.Map the types aren't very obvious, this felt easier to read.
However, this means we need to use a mutex.

All my operations require a write, so I used `sync.Mutex` instead of RWMutex but open to thoughts.